src/python/pants/backend/jvm/tasks/classpath_products.py

```diff
@@ -247,23 +249,41 @@ def get_for_targets(self, targets):
               jars.
     :rtype: list of (string, string)
     """
-    cp_entries = self.get_classpath_entries_for_targets(targets)
+    cp_entries = self.get_classpath_entries_for_targets(targets, context=context)
     return [(conf, cp_entry.path) for conf, cp_entry in cp_entries]
 
-  def get_classpath_entries_for_targets(self, targets, respect_excludes=True):
+  def get_classpath_entries_for_targets(self, targets, respect_excludes=True, context=None):
     """Gets the classpath products for the given targets.
 
     Products are returned in order, optionally respecting target excludes.
 
     :param targets: The targets to lookup classpath products for.
     :param bool respect_excludes: `True` to respect excludes; `False` to ignore them.
+    :param pants.goal.context.Context context: If not `None`, will attempt to materialize any
+      classpath entries to disk which were generated by hermetic execution and which have not
+      already been materialized to disk.
     :returns: The ordered (conf, classpath entry) tuples.
     :rtype: list of (string, :class:`ClasspathEntry`)
     """
 
     # remove the duplicate, preserve the ordering.
-    return list(OrderedSet([cp for cp, target in self.get_product_target_mappings_for_targets(
-                            targets, respect_excludes)]))
+    classpath_entries = list(OrderedSet([
+      cp for cp, target in self.get_product_target_mappings_for_targets(targets, respect_excludes)
+    ]))
+
+    if context is not None:
+      missing_entries = [
+        entry for (scope, entry) in classpath_entries
+        if entry.directory_digest and not entry.has_been_materialized
+      ]
```

Comment:
Is there a reason we need `.has_been_materialized` to be holding a lock here? Unless there's some issue with data corruption when accessing booleans across threads, it seems like the check for existence in `store.rs` is all the synchronization we need?

Reply:
Python's concurrency model is that if you're accessing a unit of memory across threads, you need synchronisation. See my other comment for why we have the boolean at all, rather than just relying on the idempotence of the rust code.

Reply:
(This also addresses a minor race condition where one thread may call `open`, the other thread may see that the file exists and decide it doesn't need to materialize, but the first thread hasn't finished writing its data yet, where thread 2 would see a partial write from thread 1)